src/expr/dtype.h

```diff
@@ -0,0 +1,620 @@
+   *
+   * @param em the NodeManager at play
+   * @param resolutions a map of strings to TypeNodes currently under
+   * resolution
+   * @param placeholders the types in these DTypes under resolution that must
+   * be replaced
+   * @param replacements the corresponding replacements
+   * @param paramTypes the sort constructors in these DTypes under resolution
+   * that must be replaced
+   * @param paramReplacements the corresponding (parametric) TypeNodes
+   */
+  bool resolve(const std::map<std::string, TypeNode>& resolutions,
+               const std::vector<TypeNode>& placeholders,
+               const std::vector<TypeNode>& replacements,
+               const std::vector<TypeNode>& paramTypes,
+               const std::vector<TypeNode>& paramReplacements);
+
+  /** compute the cardinality of this datatype */
+  Cardinality computeCardinality(TypeNode t,
+                                 std::vector<TypeNode>& processing) const;
+  /** compute whether this datatype is a recursive singleton */
+  bool computeCardinalityRecSingleton(TypeNode t,
+                                      std::vector<TypeNode>& processing,
+                                      std::vector<TypeNode>& u_assume) const;
+  /** compute whether this datatype is well-founded */
+  bool computeWellFounded(std::vector<TypeNode>& processing) const;
+  /** compute ground term
+   *
+   * This method checks if there is a term of this datatype whose type is t
+   * that is finitely constructable. As needed, it traverses its subfield types.
+   *
+   * The argument processing is the set of datatype types we are currently
+   * traversing.
+   *
+   * The argument isValue is whether we are constructing a constant value. If
+   * this flag is false, we are constructing a canonical ground term that is
+   * not necessarily constant.
+   */
+  Node computeGroundTerm(TypeNode t,
+                         std::vector<TypeNode>& processing,
+                         bool isValue) const;
+  /** Get the shared selector
+   *
+   * This returns the index^th (constructor-agnostic)
+   * selector for type t. The type dtt is the datatype
+   * type whose datatype is this class, where this may
+   * be an instantiated parametric datatype.
+   *
+   * In the terminology of "DTypes with Shared Selectors",
+   * this returns the term sel_{dtt}^{t,index}.
+   */
+  Node getSharedSelector(TypeNode dtt, TypeNode t, unsigned index) const;
+  /**
+   * Helper for mkGroundTerm and mkGroundValue above.
+   */
+  Node mkGroundTermInternal(TypeNode t, bool isValue) const;
+}; /* class DType */
+
+/**
+ * A hash function for DTypes.  Needed to store them in hash sets
+ * and hash maps.
+ */
+struct DTypeHashFunction
+{
+  size_t operator()(const DType& dt) const
+  {
+    return std::hash<std::string>()(dt.getName());
+  }
+  size_t operator()(const DType* dt) const
+  {
+    return std::hash<std::string>()(dt->getName());
+  }
+  size_t operator()(const DTypeConstructor& dtc) const
```

Comment:
Wouldn't it make more sense to define the hash functions for `DTypeConstructor` and `DTypeConstructor*` in a separate struct that lives in `dtype_cons.h`?

Reply:
Sure, this was copied over from datatype.h. I'm not sure how important these hash functions are anyways. I don't think we ever hash on datatypes.